src/core/xref.js

```diff
@@ -42,18 +42,32 @@ class XRef {
     this._cacheMap = new Map(); // Prepare the XRef cache.
     this._pendingRefs = new RefSet();
     this.stats = new DocStats(pdfManager.msgHandler);
-    this._newRefNum = null;
+    this._newRefNumForPrinting = null;
+    this._newRefNumForSaving = null;
   }
 
-  getNewRef() {
-    if (this._newRefNum === null) {
-      this._newRefNum = this.entries.length || 1;
+  getNewRefForPrinting() {
```

Comment:
Do we actually need two *slightly different* ways of generating new Refs, since that feels somewhat error-prone?
In any case, isn't there potentially a risk that if the circumstances are "just right" these methods could return the same Ref and thus break the document?

Reply:
When saving, we use some new ref only to write them but we never consume them: I mean we never get the object associated with them.
Before the patch, we were only using some new refs in the saving context.
Now, the situation is a bit different:
 - when we've a `FreeText` with no appearance, we create a `FakeUnicodeFont` which uses some references and these one must be persistent;
 - when printing we can create some `FakeUnicodeFont` which don't have to live once the printing is done, but the printing is done async, so must take to have persistent refs at least during all the printing session;
 - when saving, it's the same as before: so the refs don't have to be persistent after the saving is done.

When saving we just have to add some new ref after the last existing ref (in the pdf) whatever the number of FreeText annotations without AP or the number of time the document has been printed: it's why I think we must have something specific for saving.

In the current state of this patch, nothing is done to cleanup the refs added during the printing: it's a pity we should do something, but we must take care to not cleanup the refs added for the FreeText without AP.

Anyway, I'm not super happy with the new method names, and I agree that the implementation is not ideal.
Would you have any suggestions ?

Reply:
Good overview, thank you!

Perhaps better method names would indeed be helpful as a start, so how about something like:
 - `getNewPermanentRef` - for use with printing.
 - `getNewTemporaryRef` - for use with saving.
 - `resetNewTemporaryRef`